File: gherkin/java/src/main/java/gherkin/GherkinDialect.java
package gherkin;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@SuppressWarnings("unchecked")
public class GherkinDialect {
    private final Map<String, Object> keywords;
    private String language;

    public GherkinDialect(String language, Map<String, Object> keywords) {
        this.language = language;
        this.keywords = keywords;
    }

    public List<String> getFeatureKeywords() {
        return (List<String>) keywords.get("feature");
    }

    public String getName() {
        return (String) keywords.get("name");
    }
    public String getNativeName() {
        return (String) keywords.get("native");
    }

    public List<String> getScenarioKeywords() {
        return (List<String>) keywords.get("scenario");
    }

    public List<String> getStepKeywords() {
        List<String> result = new ArrayList<>();
        result.addAll(getGivenKeywords());
        result.addAll(getWhenKeywords());
        result.addAll(getThenKeywords());
        result.addAll(getAndKeywords());
        result.addAll(getButKeywords());
        return result;
    }

    public List<String> getBackgroundKeywords() {
        return (List<String>) keywords.get("background");
    }

    public List<String> getScenarioOutlineKeywords() {
        return (List<String>) keywords.get("scenarioOutline");
    }

    public List<String> getExamplesKeywords() {
        return (List<String>) keywords.get("examples");
    }

    public List<String> getGivenKeywords() {
        return (List<String>) keywords.get("given");
    }

    public List<String> getWhenKeywords() {
        return (List<String>) keywords.get("when");
    }

    public List<String> getThenKeywords() {
        return (List<String>) keywords.get("then");
    }

    public List<String> getAndKeywords() {
        return (List<String>) keywords.get("and");
    }

    public List<String> getButKeywords() {
        return (List<String>) keywords.get("but");
    }

    public String getLanguage() {
        return language;
    }
}

File: gherkin/php/src/GherkinDialect.php
<?php

declare(strict_types=1);

namespace Cucumber\Gherkin;

/**
 * @psalm-type Dialect = array{
 *     feature: non-empty-list<non-empty-string>,
 *     background: non-empty-list<non-empty-string>,
 *     scenario: non-empty-list<non-empty-string>,
 *     scenarioOutline: non-empty-list<non-empty-string>,
 *     examples: non-empty-list<non-empty-string>,
 *     rule: non-empty-list<non-empty-string>,
 *     given: non-empty-list<non-empty-string>,
 *     when: non-empty-list<non-empty-string>,
 *     then: non-empty-list<non-empty-string>,
 *     and: non-empty-list<non-empty-string>,
 *     but: non-empty-list<non-empty-string>,
 * } $data
 */
final class GherkinDialect
{
    /**
     * @param Dialect $dialect
     */
    public function __construct(
        private readonly string $language,
        private readonly array $dialect,
    ) {
    }

    public function getLanguage(): string
    {
        return $this->language;
    }

    /** @return non-empty-list<non-empty-string> */
    public function getFeatureKeywords(): array
    {
        return $this->dialect['feature'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getBackgroundKeywords(): array
    {
        return $this->dialect['background'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getScenarioKeywords(): array
    {
        return $this->dialect['scenario'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getScenarioOutlineKeywords(): array
    {
        return $this->dialect['scenarioOutline'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getRuleKeywords(): array
    {
        return $this->dialect['rule'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getGivenKeywords(): array
    {
        return $this->dialect['given'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getWhenKeywords(): array
    {
        return $this->dialect['when'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getThenKeywords(): array
    {
        return $this->dialect['then'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getAndKeywords(): array
    {
        return $this->dialect['and'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getButKeywords(): array
    {
        return $this->dialect['but'];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getStepKeywords(): array
    {
        return [
            ...$this->getGivenKeywords(),
            ...$this->getWhenKeywords(),
            ...$this->getThenKeywords(),
            ...$this->getAndKeywords(),
            ...$this->getButKeywords(),
        ];
    }

    /** @return non-empty-list<non-empty-string> */
    public function getExamplesKeywords(): array
    {
        return $this->dialect['examples'];
    }
}
